R/layout.R

```diff
@@ -162,9 +162,39 @@ config <- function(p, ..., cloud = FALSE, showSendToCloud = cloud, locale = NULL
   
   args <- list(...)
   if ("collaborate" %in% names(args)) warning("The collaborate button is no longer supported")
+  if ("modeBarButtonsToKeep" %in% names(args)) args <- modebar_keep(args)
   p$x$config <- modify_list(p$x$config, args)
   if (cloud) warning("The `cloud` argument is deprecated. Use `showSendToCloud` instead.")
   p$x$config$showSendToCloud <- showSendToCloud
 
   p
 }
+
+# Add all mode bar names to modeBarButtonsToRemove except the 
+# ones given by modeBarButtonsToKeep
+modebar_keep <- function(args) {
+  
+  modebar_names <- modebar_names()
+  x <- setdiff(modebar_names, unlist(args[["modeBarButtonsToKeep"]]))
+  args[["modeBarButtonsToRemove"]] <- as.list(x)
+  # Remove modeBarButtonsToKeep from args to prevent warning
+  args[["modeBarButtonsToKeep"]] <- NULL
+  args
+}
+
+# List of mode bar names from
+# https://github.com/plotly/plotly.js/blob/master/src/components/modebar/buttons.js
```

Comment:
It'd be cool if we could modify the [build script](https://github.com/ropensci/plotly/blob/master/inst/plotlyjs.R) I use when updating plotly.js to programmatically update this object. I imagine you might be use something like the V8 package to source this file and get the button names as an R object?

Reply:
Yep. Sure. To get the list of modebar names I simply did some kind of text-mining or scraping like so:

```
get_modebar_names <- function(url = "https://raw.githubusercontent.com/plotly/plotly.js/master/src/components/modebar/buttons.js") {
  modebar_names <- readLines(url)
  modebar_names <- stringr::str_trim(modebar_names)
  modebar_names <- modebar_names[stringr::str_detect(modebar_names, "^modeBarButtons\\.")]
  stringr::str_match(modebar_names, "modeBarButtons\\.(\\w+)")[, 2]
}
modebar_names <- get_modebar_names()
```
If this is fine for you I can add it to the build script. In that case I would rewrite it using base R functions to get rid of the `stringr `dependency.

Reply:
I would prefer to fo the V8 route if it's possible, but if it's not possible or too much trouble, we can go this route. Also, we should make it part of the the build script (as I linked to above), and write some sensible unit tests to check the value of `modebar_names` (i.e., it's a character vector that contains at least the default modebar buttons like `zoom2d`, etc)